[jvm-packages]Adds the ability to pass multiple custom evaluation

Helw150 opened the pull request:
Putting this here to gain access to the CI. This PR is based off updating and old fork of XGBoost which added this functionality to the current master state, so may have errors from merge conflict.

Helw150 commented:
Additionally, I ran ScalaFmt and google-java-fmt - which passed the XGBoost official style checks, but I am not sure if those meet the standards of XGBoost open source in practice.

hcho3 commented:
This is great. Does your fork also have watchlist feature for XGBoost4J-Spark?

CodingCat commented:
would you please avoid changing format in the same PR, though there are some style problems especially for Java code, 

we may focus on your watch list thing in this PR

Helw150 commented:
@CodingCat I just styled with `google-java-format` in order to pass the linter as quickly as possible rather than doing manual nits. I can try to revert as many style changes as possible if you'd like.

Helw150 commented:
Alternatively, let me know if you guys have a preferred formatter or a more specific ruleset of your linter I should follow. I try to leave as much as possible to linters and formatters style wise.

CodingCat commented:
I don't think the current linter is checking jvm/* java has no style checker now...scala is using scala-style-check plugin....

Helw150 commented:
Do you guys have a style guide then? I'm working on this as a side project, so haven't particularly set up a good dev environment for iterating on this so anything that simplifies my process to follow y'alls ruleset is much appreciated!

Helw150 commented:
The java did seem to be linting with Checkstyle(that was my motivation for using a formatter), so I just adhered to the linting rules Checkstyle imposed.

CodingCat commented:
what's the error you received? if you don't change any format, I believe it will not break with stylistic errors (as all the other PRs are successfully passing tests)

CodingCat commented:
our scala style is essentially a subset (removed some spark related thing) of https://github.com/databricks/scala-style-guide

Helw150 commented:
@CodingCat 
```
[ERROR] src/main/java/ml/dmlc/xgboost4j/java/XGBoost.java:[108] (sizes) LineLength: Line is longer than 100 characters (found 114).
[ERROR] src/main/java/ml/dmlc/xgboost4j/java/XGBoost.java:[139] (indentation) Indentation: 'method def' child have incorrect indentation level 6, expected level should be 4.
[ERROR] src/main/java/ml/dmlc/xgboost4j/java/XGBoost.java:[139] (sizes) LineLength: Line is longer than 100 characters (found 119).
[ERROR] src/main/java/ml/dmlc/xgboost4j/java/XGBoost.java:[207] (indentation) Indentation: 'for' child have incorrect indentation level 18, expected level should be 12.
[ERROR] src/main/java/ml/dmlc/xgboost4j/java/XGBoost.java:[207] (indentation) Indentation: 'member def type' have incorrect indentation level 18, expected level should be 12.
```

I'll just fix these manually. Running the Google Style guide formatter fixes all these Lint issues, but your style guide in practice seems to differ from the linter.

CodingCat commented:
still confused why all the other PRs didn't touch java can pass lint checker (are you referring to Travis CI's lint checker), but you have such an issue

can you just put stylistic issue aside and implement functionalities? or just follow what it says about style?

CodingCat commented:
maybe lint works as "check when there is any modification <b> in the directory</b>" since you add a new file there

Helw150 commented:
@CodingCat @hcho3 I don't believe the PR-merge CI errors stem from my changes as it stems from a C error and seems to be affecting a few other open PR's. Am I correct in that?

hcho3 commented:
@Helw150 Correct, I'm currently looking into it.

Helw150 commented:
@CodingCat You seem to be the one to look to for review on this one? Whenever you get a chance to review, I'd be super grateful :)

Helw150 commented:
Ping @CodingCat or @hcho3 for review :)

CodingCat commented:
@Helw150 we are focusing on 0.8 release, will look at it later

Helw150 commented:
Great, just making sure it made it on your radar. I'll add change `WIP`->`Ready For Reminder`. Thanks for the help

Helw150 commented:
@CodingCat @hcho3 Just a note, Next week is the last week of the internship where I have been working on adding this functionality as a side project. I would really appreciate it if one of you has a bit of time to review.

Sorry to ping again 😓

CodingCat commented:
Sorry, No guarantee on that, but we will try....one thing to note is that none of us work on this as full time, instead, we spent time which we squeeze from our daily jobs

Helw150 commented:
Yup, totally understand as it's the same for me on this PR. Just wanted to make sure it was still on your radar as I had seen new PR's get reviewed and merged as I know it's easy to forget things when they aren't on your roadmap!

If it can't get reviewed, no worries. It's living happily on our fork internally, so for our purposes no harm no foul if we can't get this merged.

hcho3 commented:
@Helw150 We do aim to get all pull requests merged eventually, so that the list doesn't exceed on page.

@CodingCat Wish I knew a bit about Scala so that I can help you here. The Java land is still quite new to me hehe

Helw150 commented:
@hcho3 No worries. I'm in no huge rush as long as it's on your list 👍

CodingCat commented on jvm-packages/xgboost4j/src/main/java/ml/dmlc/xgboost4j/java/XGBoost.java:
```diff
@@ -187,11 +208,16 @@ public static Booster train(
+        String evalInfo = "";
+        if (evals != null && evals.length > 0) {
+          for (int i = 0; i < evals.length; i++) {
+            String evalLine = booster.evalSet(evalMats, evalNames, evals[i], metricsOut);
```
isn't it replacing metricsOut for every time? 

and metricsOut is used in early stopping, etc. which makes it problematic

Helw150 replied:
Hmm, yeah. For early stopping, I'd probably need to add either an assumption about which eval metric is used, or a way for the user to choose which one matters. 

Are you fine with the first metric in the array of evaluations being used as the early stopping metric?

Helw150 replied:
Ok, disabled early stopping for multiple eval functions. I chose to simply print via rabit and continue training, but let me know if you'd prefer something else (namely a full error to be thrown).

CodingCat commented on jvm-packages/xgboost4j/src/main/scala/ml/dmlc/xgboost4j/scala/XGBoost.scala:
```diff
@@ -18,14 +18,57 @@ package ml.dmlc.xgboost4j.scala
+  XGBoostError,
+  BoosterResults,
+  IEvaluation
+}
```
unnecessary multi-lining

CodingCat commented on jvm-packages/xgboost4j/src/main/scala/ml/dmlc/xgboost4j/scala/XGBoost.scala:
```diff
@@ -18,14 +18,57 @@ package ml.dmlc.xgboost4j.scala
+      evals: Array[IEvaluation] = null,
+      earlyStoppingRound: Int = 0,
+      booster: Booster = null
+  ): BoosterResults = {
```
consolidate line 45 and 46

CodingCat commented on jvm-packages/xgboost4j/src/main/scala/ml/dmlc/xgboost4j/scala/XGBoost.scala:
```diff
@@ -18,14 +18,57 @@ package ml.dmlc.xgboost4j.scala
+      params
+        .filter(_._2 != null)
+        .mapValues(_.toString.asInstanceOf[AnyRef])
+        .asJava,
```
it's better to have a new variable for filtered params, to make code cleaner

CodingCat commented on jvm-packages/xgboost4j/src/main/scala/ml/dmlc/xgboost4j/scala/XGBoost.scala:
```diff
@@ -18,14 +18,57 @@ package ml.dmlc.xgboost4j.scala
+      obj,
+      evals,
+      earlyStoppingRound,
+      jBooster
```
I don't think you need make one parameter per line

CodingCat commented on jvm-packages/xgboost4j/src/main/java/ml/dmlc/xgboost4j/java/XGBoost.java:
```diff
@@ -136,12 +137,32 @@ public static Booster train(
           IEvaluation eval,
           int earlyStoppingRound,
           Booster booster) throws XGBoostError {
+    if(eval != null) {
```
space after `if`

CodingCat commented on jvm-packages/xgboost4j/src/main/java/ml/dmlc/xgboost4j/java/XGBoost.java:
```diff
@@ -136,12 +137,32 @@ public static Booster train(
+      IObjective obj,
+      IEvaluation[] evals,
+      int earlyStoppingRound,
+      Booster booster) throws XGBoostError {
```
it looks like you have empty evals and null both indicate no customized evaluation function

to make the logic clear, we'd make only one of them to be a valid input and the other triggers an exception

Helw150 replied:
If there are no evaluation functions, I chose to allow null and throw error on empty array of evaluation fucntions.

CodingCat commented on jvm-packages/xgboost4j/src/main/scala/ml/dmlc/xgboost4j/scala/XGBoost.scala:
```diff
@@ -52,22 +95,80 @@ object XGBoost {
+      obj,
+      evals,
+      earlyStoppingRound,
+      booster
```
unnecessary multi-lining

CodingCat commented on jvm-packages/xgboost4j/src/main/scala/ml/dmlc/xgboost4j/scala/XGBoost.scala:
```diff
@@ -52,22 +95,80 @@ object XGBoost {
       metrics: Array[Array[Float]] = null,
       obj: ObjectiveTrait = null,
-      eval: EvalTrait = null,
+      eval: IEvaluation = null,
```
EvalTrait is not used at all?

CodingCat commented on jvm-packages/xgboost4j/src/main/scala/ml/dmlc/xgboost4j/scala/XGBoost.scala:
```diff
@@ -94,11 +195,20 @@ object XGBoost {
       obj: ObjectiveTrait = null,
-      eval: EvalTrait = null): Array[String] = {
+      eval: EvalTrait = null
+  ): Array[String] = {
```
consolidate 198 and 199

CodingCat commented on jvm-packages/xgboost4j/src/main/scala/ml/dmlc/xgboost4j/scala/XGBoost.scala:
```diff
@@ -52,22 +95,80 @@ object XGBoost {
+      obj,
+      evals,
+      earlyStoppingRound,
+      booster
```
unnecessary multi-lining

CodingCat commented on jvm-packages/xgboost4j/src/main/scala/ml/dmlc/xgboost4j/scala/XGBoost.scala:
```diff
@@ -52,22 +95,80 @@ object XGBoost {
+      evals: Array[IEvaluation] = null,
+      earlyStoppingRound: Int = 0,
+      booster: Booster = null
+  ): Booster = {
```
consolidate 157 and 158

CodingCat commented:
any update on this?

Helw150 commented:
@CodingCat As I said a bit ago, the internship I was working on this as a side project for ended so I don't have nearly as much time to make fixes for the next couple of months. I'll try to make incremental changes based on your comments over the next few weeks, but I don't have any explicit time on my calendar to work on this while at school.

hcho3 commented:
Please ignore the failed test `continuous-integration/jenkins/pr-head`. (This was due to misconfiguration of Jenkins CI server.) Only `continuous-integration/jenkins/pr-merge` is relevant.

Helw150 commented:
@CodingCat This should be in a state for you to give another look. I fixed all the issues you had, the primary thing I am uncertain on is how I should handle the one of the lines you said shouldn't have one argument per line. 

The [style guide](https://github.com/databricks/scala-style-guide#spacing-and-indentation) you linked earlier has the following as it's standard for long method invocations, so for now I've reverted line 54-64 to a single parameter per line to meet the style guide. 

- For method and class constructor invocations, use 2 space indentation for its parameters and put each in each line when the parameters don't fit in two lines.

  ```scala
  foo(
    someVeryLongFieldName,  // 2 space indent here
    andAnotherVeryLongFieldName,
    "this is a string",
    3.1415)

  new Bar(
    someVeryLongFieldName,  // 2 space indent here
    andAnotherVeryLongFieldName,
    "this is a string",
    3.
```

hcho3 commented:
Closing as stale
